### ISSUE TITLE
Document & clarify development model

### ISSUE THREAD

**Issue by zbeekman** [sourceryinstitute/OpenCoarrays#29]
Hi Guys @sourceryinstitute, I'd love to help out some more, but I'm finding it hard to follow the branching/release/development model. It seems the two main branches, master and branch-1.0.0 [have diverged significantly](https://github.com/sourceryinstitute/opencoarrays/network).

I assume this is intentional, and you're following some kind of "git-flow" development model?

The development model should be documented somewhere and branches described as well as more detailed procedures for authoring PRs, like which branch they should be based on, and made against.

**Comment by rouson** [comment]
Hi Zaak,

I wrote a rather lengthy reply and then accidentally hit a Safari bookmark that took me to another page and I lost everything I wrote. :(  I should get in the habit of writing responses in a separate document and then pasting them here.

I have to head home for dinner with the family but hope to reconstruct what I wrote later.

Damian

**Comment by zbeekman** [comment]
Damian,
Sorry to hear you lost your hard work. No rush, just have a few minor things that I spotted and thought needed fixing.

**Comment by rouson** [comment]
Ok, Take 2:

Your question nicely assumes that we have a branching/release/development model. :)  Your help in developing one is welcome.  Alessandro and I talk weekly so you could join one of our upcoming teleconferencesto discuss this if you like.  For now, I'll explain my thoughts and influences informally and why I'm comfortable with the divergence between the two branches. 

First, the seeming divergence is partly superficial divergence because I manually copy files from one branch to the other (almost always from the release branch to the master branch).  I do this only to avoid the complications of git merge.  Git for me is like democracy, which has been described as the worst form of government except for all the alternatives.  I think possibly some of my previous difficulties related to frequent rebasing.  My life seems so much simpler now that I've stopped rebasing.  "Hi. I'm Damian and I don't rebase." :D 

Second, the minor new features that have been added on the release branch (branch-1.0.0) have very little if any interaction with the core library.  Version 1.1.0 added scripts for building the OpenCoarrays prerequisites. Version 1.2.0 added the installation script, which will see some very significant improvements in the imminent 1.2.1 release.  It's very easy and safe to copy these sorts of new features over to the master branch, which is our pre-release 2.0.0 branch as I'll explain below.

The following have most influenced my thinking on branching, version numbering, and releasing:

1. The GCC approach (https://gcc.gnu.org/develop.html), especially the version numbering scheme they adopted with release 5.
2. Semantic Versioning (http://semver.org).
3. The Forking Workflow (https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow).

Many thanks to you for referring me to the latter page.  It took quite some time and several reads for me to fully digest it.  I'm a big fan of it but kind of in the way that I'm a big fan of Zen Buddism: I often say I'm a non-practicing Zen Buddhist. ;) I haven't disciplined myself to follow the Forking Workflow in practice, but I set up my "rouson" GitHub username to facilitate private forking of the OpenCoarrays project that is hosted on the sourceryinstitute account.  Let's see if I can get back to this practice after this next release.

Our unwritten policy has been to use x.y.z version numbers to refer to major new features (x), minor new features (y), and bug fixes (z).  Hence, the master branch is 2.0.0 (the version number is in the project statement near the top of the top-level CMakeLists.txt file and it can be obtained by typing "caf --version" after a successful installation or "./install.sh --version" without installing. I'll copy the latter capability from the branch-1.0.0 release branch to the master branch as soon as I release).  
Major new features under development on the master branch include Fortran 2015 EVENT_TYPE support and support for building with our first non-GCC compiler: Portland Group.  We expect these features to appear in our 2.0.0 release, which I imagine will roughly coincide with hte GCC 6.0.0 release, which will likel happen around April 2016.

I think that pretty much covers it.  Possibly what I've written here is the beginning of a branch/release/development model.  Possibly I should massage and reformat this into a document to add to the repository.  We could then edit and discuss it as we iterate toward a more formal policy statement.

Thoughts?

**Comment by rouson** [comment]
I just seem to have all the luck lately. As fate would have it, I started having hardware problems while conducting what I'm reasonably confident would have been my last test before releasing version 1.2.1. I might have a day or so of downtime while I attempt to back the computer up. (My computer automatically backs up wirelessly when I'm at home, but I was traveling over the holiday weekend and it appears the latest backup was a week ago).  After I back up (or if I can't back up), I'll put the computer in for service, pick up a loaner, and replace the disk image on the loaner using my back up.  I imagine I'll be up and running by late in the day tomorrow sometime. 

How does all this relate to this thread?   I'm glad you asked. :)   In order that the next 24 hours won't be complete downtime, there's a feature I'd like to work on during the above process, but I want to be sure there is no chance that I'll break something that delays the release. (I'm not likely to release until I have access to the aforementioned disk image because it contains the Linux virtual machine on which I was running the final pre-release test.)

Forking WorkFlow to the rescue!  I'll work on the new feature on my "rouson" account fork (or on a new fork if merging the main repository into the existing "rouson" fork doesn't go smoothly).  Then I can safely keep everything I do on this new feature out of the main repository in case I decide to trash it.  That's what I love the most about Forking WorkFlow.

:D

**Comment by rouson** [comment]
@zbeekman, I just tried to merge the branch-1.0.0 into my private rouson/opencoarrays fork.  It appears the the result was to merge the branch into the master on rouson/opencoarrays repository.  I'm not sure if that's a good idea or not. Testing will tell, but it was surprising and yet another case where merging was sufficiently problematic that it might cost me more time than just copying the files to the appropriate branch.  The most confusing part was that the GitHub GUI gave me a message indicating that it was safe to delete the branch. I wasn't sure why it was suggesting that I would want to do that.  I did it and then it gave me a button to recover the branch.  In either state -- prior to deleting the branch and after restoring it -- I get an error message indicating the the file 'branch-1.0.0' doesn't exist when I type 'git checkout branch-1.0.0'.  Oh bother...

**Comment by zbeekman** [comment]
Boy, there is a lot of content here and a lot to discuss... I only ever use the command line these days, because I find it more difficult to figure out what clients like sourcetree and githubdesktop will do. I think this relates to your most recent woes: http://stackoverflow.com/a/72156/403516 When I have time I can try to catch up on everything else in this thread.

**Comment by rouson** [comment]
Sorry.  I wasn't referring to the GUI to which I think you're referring.   I simply meant the GitHub web interface, which I do find easier than the command line because it attempts to automate merges and pull requests and gives a nice presentation of diffs.  I use the command line for everything except merges and pull requests, which I guess is redundant in the sense that pull requests are a special case of merges.  I just wanted to highlight that the web interface is especially nice for handling pull requests. 

I had a feeling that my use of the term "GUI" would be confusing. I struggle with that trade-off between brevity and clarity and too often err on the side of being overly verbose so I was trying to curb that tendency.  

D

**Comment by zbeekman** [comment]
One problem with the manual copy branch method, is it's even hard to determine what's diverged and how. If I try to compare master to branch-1.0.0 and take a look at the [install_prerequisites/build] on github I can't even see what the differences are... https://github.com/sourceryinstitute/opencoarrays/compare/master...branch-1.0.0?expand=1#diff-c3d7a108ce2265a38caf9379460b3199

**Comment by sourceryinstitute** [comment]
Nothing has changed outside of install.sh and install_prerequisites since my October 22 commit.  In both install.sh and every file in install_prerequisites, the branch-1.0.0 versions of any files can completely replace any versions on the master branch and any files missing from install_prerequisites can be deleted on the master branch.  The copy will take me seconds or minutes so it's not something to attempt yourself if you're uncertain.  It's trivial for me to do, but there was no point in doing it until everything was working, which didn't  happen until just before I released 1.2.1 last week, at which I also experienced a hardware failure from which I didn't recover until Saturday and then I had to travel on Sunday, teach all day on Monday and Tuesday and give a talk yesterday so today id the first day this could reasonably happen. I'll be glad to do make the copies when we talk tomorrow if not sooner.

**Comment by zbeekman** [comment]
Sure, I don't mean to badger you... Just trying to get oriented.

**Comment by rouson** [comment]
Cool.  Thanks for hit continued interest, contributions and thorough review. If I had more time to explain last week, I might have been able to save you some time. It would be great if you would join an upcoming call with Alessandro and me to recommend best practices around branching and merging. 

Do you agree that the forking workflow described on Atlassian's web page is the best  ways to go?  I think that's what you referred me to last year. It worked pretty well for the class I was teaching, but that was a scenario in which conflicts were nearly impossible so merging was always easy. By contrast, on other projects, I frequently run into to problems with merges and give up. 

Anecdotally, my observation is that many people, including me, develop a workflow of occasionally cloning a fresh copy of the central repository, manually copying changes over to the new clone, and abandoning the local copy of the repository that they were editing before the fresh clone because it has become too challenging to push and deal with related confusing git messages.  I'd go so far as to say this is the most common workflow that most people adopt quietly without admitting it publicly. It's kind of like everyone's dirty little secret. I hope there is a workflow that is straightforward and makes this secret practice unnecessary. I once thought that supposed workflow involved rebasing and then recently I had a conversation with someone who used to really push me to rebase and found out she had stopped rebasing secretly but never told me. I had held onto the practice much longer and had only quit rebasing just days before the conversation. 

SMH... So much of my life I lost rebasing.  I'm sure there's a street corner somewhere where dealers sell rebasing just to get you hooked and then abandon you when the crap hits the fan... Just kidding of course. :)

**Comment by zbeekman** [comment]
Sure, I think it's worth talking about and thinking about ways to enable and enhance collaboration by changing/improving the git workflow... We can discuss this on the phone and at a later date with Alessandro.

Some thoughts on rebasing, since you brought it up:

I think rebasing has a place when you are working on say, a feature branch, and want to pick up some improvements/bug fixes that take place on master... but ONLY if you have not pushed your feature branch to the central repository (i.e. the feature branch of sourceryinstitute/opencoarrays) or any other repository that other people are working from. If, in the past, you have pushed the branch to your forked version's repository (i.e. the feature branch of rouson/opencoarrays or zbeekman/opencoarrays) then you will need to rewrite history (which you did when you rebased) and force push it to your (private!) fork. In general, avoid rebasing branches that have already been pushed somewhere where other people might be working.

This is what I typically do to decide how to proceed:
Before trying to incorporate changes into a branch, I ask myself:

 - Do I want all the commits in master (or some other branch) since they diverged?
   - If I only want some of them, then I should use `git cherry-pick -x <commit-sha>` (-x adds a message to the log saying which commit you cherry-picked from) to pull in the changes from individual commits. Each commit should be a minimal set of atomic changes that advances the state of the branch from one functional state (compiles, etc.) to another, implementing a single logical change.
  - If you do want all the commits then ask yourself: Have I pushed this branch to a remote repository yet, or does it only exist locally?
    - if it only exists locally then try something like `get pull --rebase <master-repository> master`
      - If there are merge conflicts you should be able to walk through them with `git merge --continue` or `git rebase --continue` etc. as advised by `git status`. You can always abort and then fetch the upstream/origin/central repositories changes with `get fetch <repo>` and then `git merge` or `git rebase` using an appropriate merge strategy or merge strategy option. (You can resolve all conflicts to prefer yours for example with the strategy options.)
    - If you have pushed your branch to your **private** fork repository, and you're confident no one else is using it, then you can attempt a rebase as described above, so long as you make sure to propagate the changes to your private tracking remote with `git push -f <my_private_remote>`
    - If this is not the case, then you should never rebase, since you would have to overwrite common, shared history, leading to a lot of confusion. (If you did rebase and then forced a push to the common/central repository the next time anyone else tries to update their local copies all hell will break loose since the histories are now conflicting.)
      - in this case, if you want all the changes from the other branch, you should use a normal merge/pull. If you only want some changes, then use git cherry-pick.

One final note: `reset` is a pretty confusing command, but combined with `git reflog` it is a powerful way to restore your local copy after messing something up horrifically. In concept, it's not terribly different than just scrapping the local repository and cloning a new one, although it has more fine grained control to help you hang on to more of your work. It's a lengthy read, but [this](https://git-scm.com/blog) really helped clarify some things about git for me.

**Comment by zbeekman** [comment]
Let's continue the discussion from the bottom of PR #53 here.

Hi @rouson, I know you prefer video tutorials, so I just thought I would link you to GitHub's youtube channel. They have a wealth of tutorials and webcasts to help make us into true git ninjas: https://www.youtube.com/user/GitHubGuides/playlists

Also, the "github flow" workflow is my personal preference, and might be worthwhile contemplating adoption of this: https://guides.github.com/introduction/flow/ and is detailed in an hour long seminar here: https://www.youtube.com/watch?v=EwWZbyjDs9c&index=5&list=PLg7s6cbtAD17uAwaZwiykDci_q3te3CTY

As for the videos, I highly recommend watching all of the videos in the [GitHub and git foundations playlist](https://www.youtube.com/playlist?list=PLg7s6cbtAD15G8lNyoaYDuKZSKyJrgwB-), each one is pretty short, around 5 minutes.

In particular, these more advanced videos I think are *highly* informative:
 - Committing frequently, and the safety net of [reflog](https://www.youtube.com/watch?v=Vxc9m_OVyo0&list=PLg7s6cbtAD15G8lNyoaYDuKZSKyJrgwB-&index=21)
 - Using [rebase](https://www.youtube.com/watch?v=SxzjZtJwOgo&list=PLg7s6cbtAD15G8lNyoaYDuKZSKyJrgwB-&index=22) to promote clarity in git history, to enable collaboration in open source projects.
 - Restoring previous points in history or applying fixes using and the power of [reset and checkout](https://www.youtube.com/watch?v=BKPjPMVB81g&list=PLg7s6cbtAD15G8lNyoaYDuKZSKyJrgwB-&index=20)
 - Understanding [the network and git remotes](https://www.youtube.com/watch?v=xdao5LCNoYE&list=PLg7s6cbtAD15G8lNyoaYDuKZSKyJrgwB-&index=15)
 - How git understands code movement across files and preserving history even during file renames, and [how it relates to the "ship of Theseus" philosophy problem](https://www.youtube.com/watch?v=qk1UE-9qf0Q&list=PLg7s6cbtAD15G8lNyoaYDuKZSKyJrgwB-&index=10)

**Comment by zbeekman** [comment]
See also #52 for some discussion. We are moving to [Github Flow](https://guides.github.com/introduction/flow/).

**Comment by zbeekman** [comment]
Alright, this transformation has started. Next up is to add the documentation part.

 - [ ] document development workflow, procedures and philosophy
    - [ ] update CONTRIBUTING.md

**Comment by zbeekman** [comment]
I think my recent changes to CONTRIBUTING.md and other documentation have done an OK job implementing this. Closing for now.
